sympy/core/expr.py

```diff
@@ -1116,6 +1116,8 @@ def as_poly(self, *gens, **args):
                 return poly
         except PolynomialError:
             return None
+        except GeneratorsNeeded:
+            return None
```

Comment:
You can combine this:
```
except (PolynomialError, GeneratorsNeeded):
```
It would be good to add a comment here explaining what would cause either of these exceptions to be raised and why they are being caught.

If this is changing `as_poly` then there should be tests added for `as_poly` in `test_expr.py`.

Reply:
@oscarbenjamin I added the changes as requested.